core/src/apps/zcash/orchard/crypto/note_encryption.py

```diff
@@ -0,0 +1,120 @@
+import gc
+from micropython import const
+from typing import TYPE_CHECKING
+
+from trezor.crypto import chacha20poly1305
+from trezor.crypto.hashlib import blake2b
+from trezor.crypto.pallas import Point
+from trezor.utils import empty_bytearray, ensure
+
+from apps.common.writers import write_bytes_fixed
+
+from .note import Note
+
+if TYPE_CHECKING:
+    from typing import Iterable
+    from ..random import ActionShieldingRng
+
+BLOCK_SIZE = const(64)
+ENC_CIPHERTEXT_SIZE = const(580)
+OUT_CIPHERTEXT_SIZE = const(80)
+
+
+# https://zips.z.cash/protocol/nu5.pdf#concreteorchardkdf
+def kdf_orchard(shared_secret: Point, ephemeral_key: bytes) -> bytes:
+    digest = blake2b(outlen=32, personal=b"Zcash_OrchardKDF")
+    digest.update(shared_secret.to_bytes())
+    digest.update(ephemeral_key)
+    return digest.digest()
+
+
+# https://zips.z.cash/protocol/nu5.pdf#concreteprfs
+def prf_ock_orchard(
+    ovk: bytes,
+    cv: bytes,
+    cmx: bytes,
+    ephemeral_key: bytes,
+) -> bytes:
+    digest = blake2b(outlen=32, personal=b"Zcash_Orchardock")
+    digest.update(ovk)
+    digest.update(cv)
+    digest.update(cmx)
+    digest.update(ephemeral_key)
+    return digest.digest()
+
+
+def chunks(size: int, length: int) -> Iterable[tuple[int, int]]:
+    offset = 0
+    while offset + size <= length:
+        yield offset, offset + size
+        offset += size
+    if offset < length:
+        yield offset, length
+    return
+
+
+# https://zips.z.cash/protocol/nu5.pdf#concretesym
+def sym_encrypt(key: bytes, buffer: bytearray) -> None:
+    nonce = 12 * b"\x00"
+    cipher = chacha20poly1305(key, nonce)
+    for i, j in chunks(BLOCK_SIZE, len(buffer)):
+        buffer[i:j] = cipher.encrypt(buffer[i:j])
```

Comment:
and then
```suggestion
    for chunk in chunks(BLOCK_SIZE, buffer):
        chunk[:] = cipher.encrypt(chunk)
```

Reply:
commited as suggested 758d4dc2118c221e9e54e55131132c0719341be5